### ISSUE TITLE
Issue with image_select

### ISSUE THREAD

**Issue by mquestionable** [reduxframework/redux-framework#2424]
Hello,

After selecting one option from image_select (id:header_preset) and saving it, it's okay, but if I refresh, there is no more border around option I selected and if I save again, I get error on front "Undefined index: header_preset". I tested it with code from sample and I have same issue.

Is this enough for you guys or need something more from me? :
http://support.redux.io/?id=kht72H7Jh5TGz0bSH7XyQM4YRkC%2BoIXIewUp%2BMrxAHPGB4mPd5Lqm5ucTrpPX60fNLaUwLg5O%2FPI%2BD0CR2FDOA%3D%3D

I hope you understand me, Thanks.

**Comment by kprovance** [comment]
So far, I am unable to reproduce.  Check your PHP error logs, see if there is something related.  Someone else a similar issue that was related to a security setting on the server.

**Comment by mquestionable** [comment]
I'm working on localhost. And nope, not a single one error in a while.

If you want, I can record screen while doing it, if that would help you, upload it somewhere and send you a link to some your e-mail address? Or what else?

**Comment by kprovance** [comment]
No, a video won't help.  Are you building your config with the Redux API, or the legacy way?

**Comment by mquestionable** [comment]
Redux API way.

**Comment by kprovance** [comment]
Honestly, I'm at a loss here.  I can't reproduce, and I can't fix what I can't break.

**Comment by mquestionable** [comment]
I have this code in header.php:
```
$header_style = $theme_opt['header_preset'];
```
It works alright until I refresh and save again. After that, there is no that index in theme_opt variable. (you probably concluded this from error). Can you try with this code in header and do steps above I talked about?

Thank you so much for support. When I'm able, you'll get donation from me, that's for sure.

**Comment by kprovance** [comment]
The problem isn't with your code (although you can use isset to check for uninitialized variables), the problem is - for whatever reason - when you choose your option and save, the save isn;t saving, and thus the variable comes up without index.  The fact that your options are not saving is what I cannot reproduce.  I've run the demo code, I've run your config, and everything saves as it should.  Since you are on a localhost, there is no server for me to login into to see why.  Have you tried running any of this on a live server for reproduction purposes?

**Comment by mquestionable** [comment]
I'll setup tomorrow on the server so you can see it. Where should I send you login info?

**Comment by kprovance** [comment]
You didn't answer my question, have you actually set it up on a live server and replicated the issue?

**Comment by mquestionable** [comment]
Sorry. No, I didn't. I'll tomorrow test it and if I get the same issue, I'll let you know, okay?

**Comment by zuper** [comment]
There is an issue with the image_select field

<strong> PRESET MENU </strong>
<pre>
array(
                        'id'        => 'menu_preset',
                        'type'      => 'image_select',
                        'presets'   => true,
                        'title'     => __('Menu theme', 'redux-framework-demo'),
                        'subtitle'  => __('Select a menu theme appearance or create a custom one', 'redux-framework-demo'),
                        'desc'      => __('This allows you to set a json string or array to override multiple preferences in your theme.', 'redux-framework-demo'),
                        'options'   => array(
                            '0'  => array(
		                                        'alt' => 'Menu theme 1',
		                                        'img' =>  URI_SETTINGS_PATH . 'assets/images/menu_theme/menu_theme_1.png',
		                                        'presets' => array(
			                                        'menu_position'            => 'top-sticky'
		                                        )
                            ),
                            '1'  => array(
		                                        'alt' => 'Menu theme 2',
		                                        'img' => URI_SETTINGS_PATH . 'assets/images/menu_theme/menu_theme_2.png',
		                                        'presets' => array(
			                                        'menu_position'            => 'top'
		                                        )
                            )
                        ),

                        'default'   => 0,
                    )
</pre>

As you see above if the 'default' is  0 the image_select does not work. If the 'default' is 1 the image_preset selects option '1', Both options are pointing to the field 'menu_position'

Can you please explain why when using 'default'  to '1' works ? but not on '0' if both point to the same field type ?

Find bellow the field: 'menu_position'
<strong> MENU POSITION </strong>
<pre>
array(
                        'id'        => 'menu_position',
                        'type'      => 'image_select',
                        'compiler'  => true,
                        'title'     => __('Menu position', 'redux-framework-demo'),
                        'subtitle'  => __('', 'redux-framework-demo'),
                        'desc'      => 'Dummy description',
						'options'   => array(
                            'top'           => array('alt' => 'Menu Top',           'img' => URI_SETTINGS_PATH    . 'assets/images/icon_menu_top.jpg'),
                            'top-sticky'    => array('alt' => 'Menu Top Sticky',    'img' => URI_SETTINGS_PATH   . 'assets/images/icon_menu_top_sticky.jpg')
                        ),
                        'default'   => 'top'
                    ),
</pre>

**Comment by zuper** [comment]
I have clicked on reset all button. And after the reset again nothing is selected as default although option 0 is as default. If I go to the Export tab and click 'copy data' I can clearly see that the value exists -> "menu_preset":"0" 

Also something else strange i noticed. I was changing the value in the options php file and and after doing a refresh the text under the Menu theme was: 'Default: Menu theme 1' but the image was not selected.

So it reads the value( since it shows the option alt ) but doesn't select the image.

The field image_select works ok when 'presets' is false but for sure is having issues when true.

**Comment by mquestionable** [comment]
@kprovance I tested it and sadly, I replicated the issue. What should I do next?

**Comment by zuper** [comment]
I think i found why it doesn't work.

So If i reset all the settings and 'menu_position' has the value 'top' and 'menu_preset' has the option '0'( aka menu_preset = 'top-sticky' ) it doesn't select the preset.

Basically the 'menu_preset' only works correctly if the 'menu_postion' value is set as the preset.

Shouldn't the preset override the 'menu_position' and if the value from the 'preset' doesn't match the 'menu_position' the preset should have a custom preset selected ?

I have done this in one of my panels on a builder.

So basically we have 3 presets: 'theme_1' theme_2' and 'custom' when 'theme_1' is selected all the options values from the preset   override each option. If then the user changes an option that is in the current preset the preset changes to custom

**Comment by zuper** [comment]
I have looked inside the class ReduxFramework_image_select and for me achieving what i want requires small modifications on both php and js. 

Question: 
How can I load a duplicated redux field from outside redux? Basically i have duplicated the folder image_select and changed it to image_select_custom ( inside i changed php, js class and props names to image_select_custom and all files have the ending of '_custom') and I would like to load that into redux. Can i load this into redux? 

I saw https://docs.reduxframework.com/core/advanced/add-a-custom-field/  but I don't want to override the original field

**Comment by kprovance** [comment]
@zuper, do NOT piggy back other tickets.  Create a new one with a support URL.  I am NOT going to answer two people with two separate issues on the same ticket.

**Comment by zuper** [comment]
I am new here even though I have created the account a long time ago. So you could help me on this one? Could you?

**Comment by kprovance** [comment]
Not on someone elses thread, no.  Create a new ticket with a support URL.

**Comment by kprovance** [comment]
@mquestionable Please provide a support URL from the live site, and email me a login with admin privs at @info@reduxframework.com

**Comment by mquestionable** [comment]
http://support.redux.io/?id=emSkKZrVcZ3krSI4e8%2FfcJtth9GblD8HTMT1qKfxpQ%2BDbw41wPI3cTMvQc1pMRuCClsBt88U7Kyo1R4KnMZsKA%3D%3D

I'm sending you right now login info.

**Comment by kprovance** [comment]
It may be a day or so, my plate is beyond full at present.

**Comment by mquestionable** [comment]
Please, take your time. I'll wait for your response.
Thanks

**Comment by kprovance** [comment]
I looked at your site real quick.  Besides the template files, did you change ANYTHING in the Redux core?

**Comment by mquestionable** [comment]
Nothing in Redux core, I can delete and upload again ReduxCore folder of the latest version if you want.

**Comment by kprovance** [comment]
You can, but I'm wondering about the templating, if the issue may be there.  Running your config on a standard Redux install seems to work, so I wonder if the issue is there.  Try this:  Disable the templating on the live site (done by setting the template_path argument to null or blank) so that the core redux templates are in use.  See if the problem persists them, or let me know when you've disabled them and I'll check.

**Comment by mquestionable** [comment]
I disabled my custom templates. You can check it now.

**Comment by kprovance** [comment]
Okay, so...I'm not sure what's going.  I changed the `header_preset` to the bottom one and clicked Save.  I refreshed, checked the options object and the output on your index page and I see `header_preset` is set to 2.  You're not seeing the same?  That's how it works.

**Comment by mquestionable** [comment]
After you click on the one of the presets and click save, refresh admin panel and click save again (or did you did like that?). I'm still getting the same error :/

**Comment by kprovance** [comment]
Still no clue.  Zip up your theme and send it to the same email address.

**Comment by mquestionable** [comment]
Sent.

**Comment by kprovance** [comment]
Wow!  That was intense!  I found the issue (which explains why I could not reproduce), and it's an interesting find I'll need to document on our docs site.  The issue has to do entirely with the color_rgba field.  Traditionally, when setting the default, one only need set the color and the alpha, and thaat's still true, except when dealing with presets via the image_select.  When doing so, you will need to specify the rgba default in both as well.  Not to worry, you will not need to calculate rgba values, we have a helper function that does it for you.

So, to begin, each color_rgba field that you've assigned will need to be updated to look like this (this is on of your fields):

```php
    array(
        'id'       => 'header_background_color',
        'class'    => 'row_two',
        'type'     => 'color_rgba',
        'title'    => __( 'Header Background Color', 'valde' ),
        'subtitle' => __( 'If you want different background color from the current preset.', 'valde' ),
        'default'  => array(
            'color' => '',
            'alpha' => '',
            'rgba'  => ''  // ADDED DEFAULT ARG HERE
        ),
        //'output'   => array( 'body' ),
        'validate' => 'colorrgba',
),
```
See where I added the rgba arg to the default?  You will need to do this with every color_rgba field that preset uses.  Since your defaults are blank, you can set the rgba arg blank as well.  If you decidee at some future point to preset with actual color values, you'll need to do the same for the rgba argument using the helper I'll explain next.

Okay, now to fix your preset arguments in the image_select.  I only fixed the second preset in the options array (you can do the first one  :) )
```php
array(
    'id'         => 'header_preset',
    'type'       => 'image_select',
    'presets'    => true,
    'title'      => __( 'Header Presets', 'valde' ),
    'subtitle'   => __( '', 'valde' ),
    'default'    => '1',
    'options'    => array(
        '1' => array(
            'alt'     => 'Preset 1',
            'img'     => ReduxFramework::$_url . '../sample/presets/preset1.png',
            'presets'   => array(
                'header_background_color'    => array('color' => '#ffffff', 'alpha' => '1'), 
                'header_background_color_scrolled' => array('color' => '#ffffff', 'alpha' => '1'), 
                'header_menu_link_color'     => array('color' => '#999999', 'alpha' => '1'), 
                'header_menu_link_color_hover' => array('color' => '#666666', 'alpha' => '1'), 
                'header_menu_current_link_color' => array('color' => '#444444', 'alpha' => '1'), 
                'header_menu_link_top_border' => array('color' => '#FFFFFF', 'alpha' => '0'),
                'header_menu_current_link_top_border'  => array('color' => '#F64D52', 'alpha' => '1'),
                'header_menu_link_border_color' => array('color' => '#18977a', 'alpha' => '1'),
            )
        ),
        '2' => array(
            'alt'     => 'Preset 2',
            'img'     => ReduxFramework::$_url . '../sample/presets/preset2.png',
            'presets'   => array(
                'header_background_color'    => array(
                    'color' => '#252525', 
                    'alpha' => '0.5', 
                    'rgba' => Redux_Helpers::hex2rgba ( '#252525', '0.5' )),  // ADDED RGBA ARG HERE WITH HELPER.
            )
        ),
    ),
),
```
Again, you'll see where I added the the rgba argument to the array for your color picker.  We used a built in Redux helper to output the proper values for you.  it works like this (and you can use it anywhere where you might need an RGBA value)
`Redux_Helpers::hex2rgba ( $hex_value, $alpha_value ))`

So, once you've fixed up the preset values in your first options array, everything should begin to sync up.  Because the values were not matching in the image_select field code itself, the last saved choice would not set properly on refresh.  Because the option was not properly set, the next save defaulted it to blank which saved nothing to the database, which in turn gave you the index errors.

I hope this makes some sense, but if not, fixing your code as I've outlined will.  :)

**Comment by kprovance** [comment]
ps -  after this fix, you may need to reset the section to clean up the database before saving anew.

**Comment by mquestionable** [comment]
Thank you Sir and good job! Now it's (almost) finally working like it's supposed to :)

Just one more quick question (I'll open new issue ticket if I need to). I set default preset to be 1, after reset all, I get in output on index that header_preset is set to 1, but it didn't load presets. If I save options, I get undefined variable header_preset. I think this is problem like zuper has asking, is there any fix to this?
